File: eu.fbk.se.simpledg/src/main/java/eu/fbk/se/simpledg/utils/GraphAnalyzer.java
package eu.fbk.se.simpledg.utils;

import org.graphstream.graph.Edge;
import org.graphstream.graph.Graph;
import org.graphstream.graph.Node;
import org.graphstream.graph.implementations.SingleGraph;
import org.graphstream.stream.file.FileSource;
import org.graphstream.stream.file.FileSourceDOT;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.StringReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;


public class GraphAnalyzer {

    private static boolean isDirected;

    public static void main(String[] args) {
        if (args.length != 1) {
            System.err.println("Usage: java GraphStatisticsAnalyzer <path-to-dot-file>");
            System.exit(1);
        }

        String dotFileName = args[0];

        try {
            Graph graph = loadGraph(dotFileName);

            // Print statistics report
            printStatisticsReport(graph);

            // Write results to CSV
            writeResultsToCSV(dotFileName, graph);

        } catch (IOException e) {
            System.err.println("Error loading the graph from the DOT file: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static Graph loadGraph(String dotFileName) throws IOException {
        Graph graph = new SingleGraph("LoadedGraph");


        File file = new File("src/main/resources/" + dotFileName);
        String absolutePath = file.getAbsolutePath();

        FileSource fileSource = new FileSourceDOT();

        try {
            fileSource.addSink(graph);

            // Read the file content
            String fileContent = readFileContent(absolutePath);

            // Perform transformations on the file content
            String transformedContent = transformFileContent(fileContent);

            // Load the modified content into the graph
            fileSource.readAll(new StringReader(transformedContent));
            // fileSource.readAll(absolutePath);
        } catch (IOException e) {
            System.err.println("Error reading DOT file: " + e.getMessage());
            throw e; // Rethrow the exception to propagate it up
        } finally {
            fileSource.removeSink(graph);
        }

        return graph;
    }

    private static String readFileContent(String filePath) throws IOException {
        // Read the content of the file
        return new String(Files.readAllBytes(Paths.get(filePath)), StandardCharsets.UTF_8);
    }

    private static String transformFileContent(String originalContent) {
        // Replace "=>" with "->" and add a semicolon at the end of each statement
        originalContent = originalContent.replaceAll("=>", "->");

        // Replace "==" with "--" and add a semicolon at the end of each statement
        originalContent = originalContent.replaceAll("==", "--");

        // Add a semicolon at the end of each statement
        originalContent = originalContent.replaceAll("(\\s*\\w+\\s*->\\s*\\w+\\s*)(?!;)", "$1;");


        // Add a semicolon at the end of each statement
        originalContent = originalContent.replaceAll("(\\s*\\w+\\s*--\\s*\\w+\\s*)(?!;)", "$1;");

        isDirected = originalContent.contains("->");

        return originalContent;
    }


    private static void printStatisticsReport(Graph graph) {
        // Calculate and print statistics
        System.out.println("Number of nodes: " + graph.getNodeCount());
        System.out.println("Number of edges: " + graph.getEdgeCount());

        try {
            System.out.println("Is a tree: " + isTree(graph));
        } catch (RuntimeException e) {
            System.err.println("Error checking if the graph is a tree: " + e.getMessage());
        }

        try {
            if (isDirected) {
                System.out.println("Is Directed Graph: Yes");
                System.out.println("Is a DAG: " + isDAG(graph));
            } else {
                System.out.println("Is Directed Graph: No");
                System.out.println("Graph is not directed. Skipping DAG check.");
            }
        } catch (RuntimeException e) {
            System.err.println("Error checking if the graph is a DAG: " + e.getMessage());
        }

        try {
            System.out.println("Is complete: " + isComplete(graph));
        } catch (RuntimeException e) {
            System.err.println("Error checking if the graph is complete: " + e.getMessage());
        }
    }

    private static void writeResultsToCSV(String dotFileName, Graph graph) {
        try {
            // Create the output folder if it doesn't exist
            File outputFolder = new File("output");
            if (!outputFolder.exists()) {
                outputFolder.mkdir();
            }

            // Get the file name without extension
            String fileNameWithoutExtension = getFileNameWithoutExtension(dotFileName);

            // Delete existing CSV file if it exists
            Path existingCsvPath = Paths.get("output/" + fileNameWithoutExtension + "_stats.csv");
            Files.deleteIfExists(existingCsvPath);

            // Write results to CSV
            try (FileWriter writer = new FileWriter(existingCsvPath.toString())) {
                writer.write("Number of nodes: " + graph.getNodeCount() + "\n");
                writer.write("Number of edges: " + graph.getEdgeCount() + "\n");
                writer.write("Is a tree: " + isTree(graph) + "\n");
                writer.write("Is a DAG: " + isDAG(graph) + "\n");
                writer.write("Is complete: " + isComplete(graph) + "\n");

                System.out.println("Results written to CSV file.");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static String getFileNameWithoutExtension(String filePath) {
        Path path = Paths.get(filePath);
        String fileNameWithExtension = path.getFileName().toString();
        int dotIndex = fileNameWithExtension.lastIndexOf('.');
        return (dotIndex == -1) ? fileNameWithExtension : fileNameWithExtension.substring(0, dotIndex);
    }


    private static boolean isTree(Graph graph) {
        int numNodes = graph.getNodeCount();
        int numEdges = graph.getEdgeCount();

        // Property 1: Number of edges should be exactly one less than the number of nodes for a connected graph
        if (numEdges != numNodes - 1) {
            return false;
        }

        // Property 2: Check if the graph is acyclic using Depth-First Search (DFS)
        Set<Node> visitedNodes = new HashSet<>();
        Node startNode = graph.getNode(0); // Assuming the graph is connected
        return isAcyclicDFS(startNode, null, visitedNodes);
    }

    private static boolean isAcyclicDFS(Node currentNode, Node parent, Set<Node> visitedNodes) {
        if (visitedNodes.contains(currentNode)) {
            return false; // Cycle detected
        }

        visitedNodes.add(currentNode);

        Iterator<? extends Node> neighborIterator = currentNode.getNeighborNodeIterator();
        while (neighborIterator.hasNext()) {
            Node neighbor = neighborIterator.next();
            if (neighbor != parent && !isAcyclicDFS(neighbor, currentNode, visitedNodes)) {
                // Early return if a cycle is detected
                return false;
            }
        }

        return true;
    }

    private static boolean isDAG(Graph graph) {
        Set<Node> visitedNodes = new HashSet<>();
        Set<Node> currentlyVisitedNodes = new HashSet<>();

        for (Node node : graph.getNodeSet()) {
            if (!visitedNodes.contains(node)) {
                if (!isDAGDFS(node, visitedNodes, currentlyVisitedNodes)) {
                    return false; // Graph is not a DAG
                }
            }
        }

        return true; // Graph is a DAG
    }


    private static boolean isDAGDFS(Node currentNode, Set<Node> visitedNodes, Set<Node> currentlyVisitedNodes) {
        visitedNodes.add(currentNode);
        currentlyVisitedNodes.add(currentNode);

        for (Edge edge : currentNode.getLeavingEdgeSet()) {
            Node neighbor = edge.getOpposite(currentNode);

            if (!visitedNodes.contains(neighbor)) {
                if (!isDAGDFS(neighbor, visitedNodes, currentlyVisitedNodes)) {
                    return false; // Graph is not a DAG
                }
            } else if (currentlyVisitedNodes.contains(neighbor)) {
                // Detected a back edge, indicating a cycle
                return false; // Graph is not a DAG
            }
        }

        currentlyVisitedNodes.remove(currentNode);
        return true; // No cycles detected in this DFS traversal
    }

    private static boolean isComplete(Graph graph) {
        for (Node node1 : graph.getNodeSet()) {
            for (Node node2 : graph.getNodeSet()) {
                if (node1 != node2 && !hasEdgeBetween(node1, node2)) {
                    // If there is no edge from node1 to node2, the graph is not complete
                    return false;
                }
            }
        }
        return true;
    }

    private static boolean hasEdgeBetween(Node source, Node target) {
        for (Edge edge : source.getEdgeSet()) {
            if (edge.getOpposite(source) == target) {
                return true; // There is an edge from source to target
            }
        }
        return false; // No edge found from source to target
    }


}

File: eu.fbk.se.simpledg/src/eu/fbk/se/simpledg/utils/graph_analyzer.py
import argparse
from collections import defaultdict
import csv
import os

class Graph:
    """
    Class to represent a graph, supporting operations to check if it's a tree, DAG, or complete graph.
    """
    def __init__(self, is_directed):
        self.vertices = 0
        self.vertex_map = {}
        self.adj_list = defaultdict(set) if is_directed else defaultdict(set)
        self.is_directed = is_directed

    def add_vertex(self, vertex):
        if vertex not in self.vertex_map:
            self.vertex_map[vertex] = self.vertices
            self.vertices += 1

    def add_edge(self, u, v):
        if u not in self.vertex_map:
            self.add_vertex(u)
        if v not in self.vertex_map:
            self.add_vertex(v)
        u_index = self.vertex_map[u]
        v_index = self.vertex_map[v]
        self.adj_list[u_index].add(v_index)
        if not self.is_directed:
            self.adj_list[v_index].add(u_index)


    def get_number_of_edges(self):
        edges = sum(len(neighbors) for neighbors in self.adj_list.values())
        return edges if self.is_directed else edges // 2

    def is_tree(self):
        visited = set()

        def dfs(node, parent):
            if node in visited:
                return False
            visited.add(node)
            for neighbor in self.adj_list[node]:
                if neighbor != parent and not dfs(neighbor, node):
                    return False
            return True

        return self.vertices == len(visited) and dfs(next(iter(self.vertex_map.values())), -1)

    def is_dag(self):
        visited = set()
        stack = set()

        def dfs(node):
            visited.add(node)
            stack.add(node)
            for neighbor in self.adj_list[node]:
                if neighbor not in visited:
                    if not dfs(neighbor):
                        return False
                elif neighbor in stack:
                    return False
            stack.remove(node)
            return True

        for node in range(self.vertices):
            if node not in visited:
                if not dfs(node):
                    return False
        return True

    def is_complete(self):
        if any(i in self.adj_list[i] for i in self.adj_list):
        # If any node has a self-loop, it's not a complete graph.
            return False

        if self.is_directed:
        # In a directed graph, check for edges in both directions.
            for i in range(self.vertices):
                for j in range(self.vertices):
                    if i != j and (j not in self.adj_list[i] or i not in self.adj_list[j]):
                        return False
        else:
            # In an undirected graph, just check for the presence of edges.
            for i in range(self.vertices):
                for j in range(i + 1, self.vertices):
                    if j not in self.adj_list[i]:
                        return False

        return True



def load_graph(dot_file):
    """
    Loads a graph from a DOT file, with support for alternative edge notations.
    """
    graph = None
    try:
        with open(dot_file, 'r') as f:
            for line in f:
                stripped_line = line.strip().strip(';')

                if "digraph" in stripped_line:
                    graph = Graph(True)
                elif "graph" in stripped_line:
                    graph = Graph(False)
                else:
                    # Adjust delimiter for different edge types
                    if "->" in stripped_line or "=>" in stripped_line:
                        delimiter = "->" if "->" in stripped_line else "=>"
                    elif "--" in stripped_line or "==" in stripped_line:
                        delimiter = "--" if "--" in stripped_line else "=="
                    else:
                        delimiter = None

                    if delimiter:
                        vertices = stripped_line.split(delimiter)
                        if len(vertices) == 2:
                            u, v = [vertex.strip() for vertex in vertices]
                            graph.add_edge(u, v)  # Add all edges directly, including self-loops
                    else:
                        # Handle standalone nodes
                        parts = stripped_line.split()
                        if parts and all(part not in ('digraph', 'graph', '{', '}') for part in parts):
                            for part in parts:
                                graph.add_vertex(part.strip())
    except IOError as e:
        print(f"Error reading file {dot_file}: {e}")
        return None

    return graph

def generate_report(graph, graph_name, dot_file):
    """
    Generates a report for the given graph and saves it into a separate CSV file in the 'output' folder.
    Each CSV file is named based on the input file name.
    """
    report_data = {
        "Graph Name": graph_name,
        "File": dot_file,
        "Number of Nodes": graph.vertices,
        "Number of Edges": graph.get_number_of_edges(),
        "Is Tree": graph.is_tree(),
        "Is DAG": graph.is_dag(),
        "Is Complete": graph.is_complete()
    }

    # Displaying on console
    print(f"{graph_name} Report for {dot_file}:")
    for key, value in report_data.items():
        print(f"{key}: {value}")
    print()

    # Create 'output' directory if it doesn't exist
    output_dir = 'output'
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    # Extracting file name without extension and appending '_stats.csv'
    base_file_name = os.path.splitext(os.path.basename(dot_file))[0]
    csv_file_name = f"{base_file_name}_stats.csv"
    csv_file_path = os.path.join(output_dir, csv_file_name)

    # Writing to CSV in the 'output' folder
    with open(csv_file_path, 'w', newline='') as file:
        writer = csv.DictWriter(file, fieldnames=report_data.keys())
        writer.writeheader()
        writer.writerow(report_data)

    print(f"Report saved to {csv_file_path}")
    print()

def main():
    parser = argparse.ArgumentParser(description="Graph Analyzer")
    parser.add_argument("dot_file", help="Path to the DOT file")
    args = parser.parse_args()

    graph = load_graph(args.dot_file)

    if graph:
        generate_report(graph, "Graph", args.dot_file)
    else:
        print("Failed to load the graph.")

if __name__ == "__main__":
    main()
